rover-apps/bms_2022/src/BMSMonitoring.cpp

```diff
@@ -0,0 +1,130 @@
+#include "BMSMonitoring.h"
+
+#include <cmath>
+#include <map>
+
+#include "Logger.h"
+#include "mbed.h"
+
+SPI spi(SDI, SDO, CLK);  // mosi, miso, sclk
+AnalogIn currentADC(CURR_ANLG_IN);
+DigitalOut buzzer_en(BUZZER_EN);
+DigitalOut chip_select(SAMPL);
+
+BMSMonitoring::BMSMonitoring() {
+  buzzer_en.write(0);
+  chip_select.write(1);
+  spi.format(24, 3);
+  spi.frequency(1000000);
+}
+
+void BMSMonitoring::cell_monitoring() {
+  chip_select.write(0);
+
+  // TODO: recieve data from can
+  // is_cell_balancing = getRXSignal
+  // can.getRXSignalValue(HWBRIDGE::CANID::BMS, HWBRIDGE::CANSIGNAL::BMS_STATE, is_cell_balancing);
+
+  // if voltage is too low on one cell, enable balancing on that cell
+  // only way to balance is to keep all fets on so cells always have same voltage
+  int spi_message = 0b000000000000000000000000;
+  if (is_cell_balancing) {
+    spi_message = 0b111111111111111100000000;
+  }
+
+  battery_voltage = 0;
+  for (int i = 0; i < 16; i++) {
+    cell_voltages[i] = spi.write(spi_message);
+    battery_voltage += cell_voltages[i];
+    spi_message += 0b10000;
+  }
+  chip_select.write(1);
+
+  update_voltage_percent();
+}
+
+void BMSMonitoring::update_voltage_percent() {
+  if (battery_voltage >= 50.40) {
+    battery_percent_voltage = 100;
+  } else if (battery_voltage >= 49.80) {
+    battery_percent_voltage = 95;
+  } else if (battery_voltage >= 49.32) {
+    battery_percent_voltage = 90;
+  } else if (battery_voltage >= 48.96) {
+    battery_percent_voltage = 85;
+  } else if (battery_voltage >= 48.24) {
+    battery_percent_voltage = 80;
+  } else if (battery_voltage >= 47.76) {
+    battery_percent_voltage = 75;
+  } else if (battery_voltage >= 47.40) {
+    battery_percent_voltage = 70;
+  } else if (battery_voltage >= 46.92) {
+    battery_percent_voltage = 65;
+  } else if (battery_voltage >= 46.44) {
+    battery_percent_voltage = 60;
+  } else if (battery_voltage >= 46.20) {
+    battery_percent_voltage = 55;
+  } else if (battery_voltage >= 46.08) {
+    battery_percent_voltage = 50;
+  } else if (battery_voltage >= 45.84) {
+    battery_percent_voltage = 45;
+  } else if (battery_voltage >= 45.60) {
+    battery_percent_voltage = 40;
+  } else if (battery_voltage >= 45.48) {
+    battery_percent_voltage = 35;
+  } else if (battery_voltage >= 45.24) {
+    battery_percent_voltage = 30;
+  } else if (battery_voltage >= 45.00) {
+    battery_percent_voltage = 25;
+  } else if (battery_voltage >= 44.76) {
+    battery_percent_voltage = 20;
+  } else if (battery_voltage >= 44.52) {
+    battery_percent_voltage = 15;
+  } else if (battery_voltage >= 44.28) {
+    battery_percent_voltage = 10;
+  } else if (battery_voltage >= 43.32) {
+    battery_percent_voltage = 5;
+  } else {
+    battery_percent_voltage = 0;
+  }
```

Comment:
I feel like there should be a better way of doing this, what about mapping 0-100% to 43.32-50.4V? Like `battery_percent_voltage = (battery_voltage - 43.32) / (50.4 - 43.32);` and just do a catch for the corner cases of V > 50.4 and V < 43.32.

Reply:
lipo battery voltages are not linear and they aren't the most accurate for state of charge estimation. I'm following [this](https://www.promodeler.com/resize/Shared/Images/Product/Battery-6S-LiPo/LiPo-Capacity-v-Voltage.png?bw=1000&w=1000&bh=1000&h=1000) battery voltage chart.